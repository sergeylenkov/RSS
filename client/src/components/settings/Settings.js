import React from 'react';
import { connect } from 'react-redux';
import { toggleTheme } from '../../store/actions/index.js';

import styles from './Settings.module.css';

export class ConnectedSettings extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            collapse: JSON.parse(localStorage.getItem('collpaseLong')),
            days: localStorage.getItem('keepDays')
        }

        this.collapseId = 0;
        this.themeId = 1;

        this.collapseFieldRef = React.createRef();
        this.daysFieldRef = React.createRef();
        this.themeFieldRef = React.createRef();

        this.onToggleCollapse = this.onToggleCollapse.bind(this);
        this.onChangeDays = this.onChangeDays.bind(this);
        this.onToggleTheme = this.onToggleTheme.bind(this);
        this.onClick = this.onClick.bind(this);
    }

    render() {
        if (!this.props.isVisible) {
            return null;
        }

        return (
            <div className={styles.container} onClick={this.onClick}>
                <div className={styles.item}>
                    <input id={this.themeId} className={styles.checkbox} ref={this.themeFieldRef} type="checkbox" checked={this.props.isDarkTheme ? 'checked' : ''} onChange={this.onToggleTheme} />
                    <label className={styles.label} htmlFor={this.themeId}>Темная тема</label>
                </div>

                <div className={styles.item}>
                    <input id={this.collapseId} className={styles.checkbox} ref={this.collapseFieldRef} type="checkbox" checked={this.state.collapse ? 'checked' : ''} onChange={this.onToggleCollapse} />
                    <label className={styles.label} htmlFor={this.collapseId}>Сворачивать длинные посты</label>
                </div>

                <div className={styles.item}>
                    <div className={styles.label}>Удалять посты старше <input className={styles.days} ref={this.daysFieldRef} type="text" value={this.state.days} onChange={this.onChangeDays}/> дней</div>
                </div>
            </div>
        );
    }

    onToggleCollapse() {
        const collapse = this.collapseFieldRef.current.checked;

        localStorage.setItem('collpaseLong', collapse);

        this.setState({
            collapse: collapse
        });
    }

    onChangeDays() {
        const days = this.daysFieldRef.current.value;

        localStorage.setItem('keepDays', days);

        this.setState({
            days: days
        });
    }

    onToggleTheme() {
        const isDarkTheme = !this.props.isDarkTheme;

        this.props.toggleTheme(isDarkTheme);
    }

    onClick(e) {
        e.stopPropagation();
        e.nativeEvent.stopImmediatePropagation();
    }
}

/* Redux */

const mapStateToProps = state => {
    return {
        isDarkTheme: state.isDarkTheme
    };
};

const mapDispatchToProps = dispatch => {
    return {
        toggleTheme: (isDarkTheme) => dispatch(toggleTheme(isDarkTheme))
    };
};

export const Settings = connect(mapStateToProps, mapDispatchToProps)(ConnectedSettings);