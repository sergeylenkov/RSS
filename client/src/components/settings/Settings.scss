.settings {
  &__container {
    position: absolute;
    right: 0;
    top: 100%;
    display: flex;
    flex-direction: column;
    padding: 10px 20px;
    border-radius: 4px 0 4px 4px;
    background: #ffffff;
    box-shadow: 0 10px 10px 1px rgba(0, 0, 0, 0.1);
  }

  &__item {
    width: 100%;
    height: 40px;
    display: flex;
  }

  &__label {
    margin: auto 0;
    font-size: 14px;
  }

  &__checkbox {
    margin: auto 10px auto 0;
  }

  &__days {
    width: 30px;
    margin: 0 5px;
    text-align: right;
  }
}
