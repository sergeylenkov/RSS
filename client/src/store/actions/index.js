import {
  CHANGE_VIEW_MODE,
  ENTRIES_LOADED,
  ENTRIES_UPDATE_ERROR,
  ENTRIES_UPDATING,
  FEEDS_ADD,
  FEEDS_DELETE,
  FEEDS_EDITING,
  FEEDS_SELECT,
  FEEDS_UPDATE,
  FEEDS_UPDATED,
  FEEDS_UPDATING,
  TOGGLE_THEME,
  UPDATE_ENTRIES_COUNT,
  UPDATE_FAVORITE,
  UPDATE_READ,
  UPDATE_UNVIEWED_COUNT,
  UPDATE_VIEWED,
  TOGGLE_COLLAPSE_LONG,
  UPDATE_KEEP_DAYS
} from '../constants/index.js';

export function feedsUpdating() {
  return { type: FEEDS_UPDATING }
}

export function feedsUpdated(feeds) {
  return { type: FEEDS_UPDATED, feeds }
}

export function feedsAdd(feed) {
  return { type: FEEDS_ADD, feed }
}

export function feedsDelete(id) {
  return { type: FEEDS_DELETE, id }
}

export function feedsEditing(isEditing) {
  return { type: FEEDS_EDITING, isEditing }
}

export function feedsSelect(id) {
  return { type: FEEDS_SELECT, id }
}

export function entriesUpdating() {
  return { type: ENTRIES_UPDATING }
}

export function entriesUpdated(entries) {
  return { type: ENTRIES_LOADED, entries }
}

export function updateUnviewedCount() {
  return { type: UPDATE_UNVIEWED_COUNT }
}

export function updateViewed(ids) {
  return { type: UPDATE_VIEWED, ids }
}

export function updateFavorite(id, isFavorite) {
  return { type: UPDATE_FAVORITE, id: id, isFavorite: isFavorite }
}

export function updateRead(id, isRead) {
  return { type: UPDATE_READ, id: id, isRead: isRead }
}

export function changeViewMode(mode) {
  return { type: CHANGE_VIEW_MODE, mode: mode }
}

export function updateEntriesCount() {
  return { type: UPDATE_ENTRIES_COUNT }
}

export function feedsUpdate(id, data) {
  return { type: FEEDS_UPDATE, id: id, data: data }
}

export function toggleTheme(isDarkTheme) {
  return { type: TOGGLE_THEME, isDarkTheme: isDarkTheme }
}

export function entriesUpdateError() {
  return { type: ENTRIES_UPDATE_ERROR }
}

export function toggleCollapseLong(isCollapse) {
  return { type: TOGGLE_COLLAPSE_LONG, isCollapse: isCollapse }
}

export function updateKeepDays(days) {
  return { type: UPDATE_KEEP_DAYS, days: days};
}
