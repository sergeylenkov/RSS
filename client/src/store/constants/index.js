export const FEEDS_UPDATING = 'FEEDS_UPDATING';
export const FEEDS_UPDATED = 'FEEDS_UPDATED';
export const ENTRIES_UPDATING = 'ENTRIES_UPDATING';
export const ENTRIES_UPDATED = 'ENTRIES_UPDATED';
export const UPDATE_UNVIEWED_COUNT = 'UPDATE_UNVIEWED_COUNT';
export const UPDATE_VIEWED = 'UPDATE_VIEWED';