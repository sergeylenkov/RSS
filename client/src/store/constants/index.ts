export const FEEDS_UPDATING = 'FEEDS_UPDATING';
export const FEEDS_UPDATED = 'FEEDS_UPDATED';
export const FEEDS_ADD = 'FEEDS_ADD';
export const FEEDS_DELETE = 'FEEDS_DELETE';
export const FEEDS_UPDATE = 'FEEDS_UPDATE';
export const FEEDS_EDITING = 'FEEDS_EDITING';
export const FEEDS_SELECT = 'FEEDS_SELECT';
export const ENTRIES_UPDATING = 'ENTRIES_UPDATING';
export const ENTRIES_LOADED = 'ENTRIES_LOADED';
export const UPDATE_UNVIEWED_COUNT = 'UPDATE_UNVIEWED_COUNT';
export const UPDATE_VIEWED = 'UPDATE_VIEWED';
export const UPDATE_FAVORITE = 'UPDATE_FAVORITE';
export const UPDATE_READ = 'UPDATE_READ';
export const CHANGE_VIEW_MODE = 'CHANGE_VIEW_MODE';
export const UPDATE_ENTRIES_COUNT = 'UPDATE_ENTRIES_COUNT';
export const TOGGLE_THEME = 'TOGGLE_THEME';
export const ENTRIES_UPDATE_ERROR = 'ENTRIES_UPDATE_ERROR';
export const TOGGLE_COLLAPSE_LONG = 'TOGGLE_COLLAPSE_LONG';
export const UPDATE_KEEP_DAYS = 'UPDATE_KEEP_DAYS';
