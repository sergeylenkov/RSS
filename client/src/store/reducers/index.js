import {
    FEEDS_UPDATING, FEEDS_UPDATED, FEEDS_ADD, FEEDS_DELETE,
    ENTRIES_UPDATING, ENTRIES_UPDATED, UPDATE_UNVIEWED_COUNT, UPDATE_VIEWED,
    UPDATE_FAVORITE, UPDATE_READ
} from '../constants/index';

const initialState = {
    isUpdating: false,
    unviewedCount: 0,
    feeds: [],
    entries: []
};

function rootReducer(state = initialState, action) {
    if (action.type === FEEDS_UPDATING) {
        return {
            ...state,
            isUpdating: true
        }
    }

    if (action.type === FEEDS_UPDATED) {
        return {
            ...state,
            isUpdating: false,
            feeds: [...action.feeds]
        }
    }

    if (action.type === FEEDS_ADD) {
        const feeds = [...state.feeds];

        feeds.push(action.feed);

        return {
            ...state,
            feeds: feeds
        }
    }

    if (action.type === FEEDS_DELETE) {
        const feeds = state.feeds.filter(feed => {
            return feed.id !== action.id
        });        

        return {
            ...state,
            feeds: feeds
        }
    }

    if (action.type === ENTRIES_UPDATING) {
        return {
            ...state,
            isUpdating: true
        }
    }

    if (action.type === ENTRIES_UPDATED) {
        return {
            ...state,
            isUpdating: false,
            entries: [...action.entries]
        }
    }

    if (action.type === UPDATE_UNVIEWED_COUNT) {
        const feeds = [...state.feeds];
        let totalCount = 0;

        feeds.forEach(feed => {
            const count = state.entries.filter(entry => {
                return entry.feedId === feed.id && entry.isViewed !== true
            }).length;

            feed.count = count;
            totalCount = totalCount + count;
        });
        
        return {
            ...state,
            feeds: feeds,
            unviewedCount: totalCount
        }
    }

    if (action.type === UPDATE_VIEWED) {
        const entries = state.entries.map(entry => {
            if (action.ids.indexOf(entry.id) !== -1) {                
                return { ...entry, isViewed: true }
            }

            return entry;
        });

        return {
            ...state,
            entries: entries
        }
    }

    if (action.type === UPDATE_FAVORITE) {
        const entries = state.entries.map(entry => {
            return entry.id === action.id ? { ...entry, isFavorite: action.isFavorite } : entry
        });

        return {
            ...state,
            entries: entries
        }
    }

    if (action.type === UPDATE_READ) {
        const entries = state.entries.map(entry => {
            return entry.id === action.id ? { ...entry, isRead: true } : entry
        });

        return {
            ...state,
            entries: entries
        }
    }

    return state;
};

export default rootReducer;