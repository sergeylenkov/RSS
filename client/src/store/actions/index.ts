import { ActionTypes } from '../constants';

export function feedsUpdating() {
  return { type: ActionTypes.FEEDS_UPDATING }
}

export function feedsUpdated(feeds: any[]) {
  return { type: ActionTypes.FEEDS_UPDATED, feeds }
}

export function feedsAdd(feed: any) {
  return { type: ActionTypes.FEEDS_ADD, feed }
}

export function feedsDelete(id: number) {
  return { type: ActionTypes.FEEDS_DELETE, id }
}

export function feedsEditing(isEditing: boolean) {
  return { type: ActionTypes.FEEDS_EDITING, isEditing }
}

export function feedsSelect(id: number) {
  return { type: ActionTypes.FEEDS_SELECT, id }
}

export function entriesUpdating() {
  return { type: ActionTypes.ENTRIES_UPDATING }
}

export function entriesUpdated(entries: any[]) {
  return { type: ActionTypes.ENTRIES_LOADED, entries }
}

export function updateUnviewedCount() {
  return { type: ActionTypes.UPDATE_UNVIEWED_COUNT }
}

export function updateViewed(ids: number[]) {
  return { type: ActionTypes.UPDATE_VIEWED, ids }
}

export function updateFavorite(id: number, isFavorite: boolean) {
  return { type: ActionTypes.UPDATE_FAVORITE, id: id, isFavorite: isFavorite }
}

export function updateRead(id: number, isRead: boolean) {
  return { type: ActionTypes.UPDATE_READ, id: id, isRead: isRead }
}

export function changeViewMode(mode: number) {
  return { type: ActionTypes.CHANGE_VIEW_MODE, mode: mode }
}

export function updateEntriesCount() {
  return { type: ActionTypes.UPDATE_ENTRIES_COUNT }
}

export function feedsUpdate(id: number, data: any) {
  return { type: ActionTypes.FEEDS_UPDATE, id: id, data: data }
}

export function toggleTheme(isDarkTheme: boolean) {
  return { type: ActionTypes.TOGGLE_THEME, isDarkTheme: isDarkTheme }
}

export function entriesUpdateError() {
  return { type: ActionTypes.ENTRIES_UPDATE_ERROR }
}

export function toggleCollapseLong(isCollapse: boolean) {
  return { type: ActionTypes.TOGGLE_COLLAPSE_LONG, isCollapse: isCollapse }
}

export function updateKeepDays(days: number) {
  return { type: ActionTypes.UPDATE_KEEP_DAYS, days: days};
}
