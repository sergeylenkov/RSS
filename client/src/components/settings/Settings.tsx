import React, { MouseEvent, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toggleTheme, toggleCollapseLong, updateKeepDays, toggleGrid } from '../../store/actions';
import { State } from '../../store/reducers';
import { Bem } from '../../utils/bem';

import './Settings.scss';

const block = new Bem('settings');

interface SettingsProps {
  isVisible: boolean;
}

function Settings({ isVisible }: SettingsProps): JSX.Element | null {
  const collapseFieldRef = useRef<HTMLInputElement>(null);
  const daysFieldRef = useRef<HTMLInputElement>(null);
  const themeFieldRef = useRef<HTMLInputElement>(null);
  const gridFieldRef = useRef<HTMLInputElement>(null);
  const isDarkTheme = useSelector<State, boolean>(state => state.isDarkTheme);
  const isCollapseLong = useSelector<State, boolean>(state => state.isCollapseLong);
  const isGrid = useSelector<State, boolean>(state => state.isGrid);
  const keepDays = useSelector<State, number>(state => state.keepDays);
  const dispatch = useDispatch();
  const collapseId = '0';
  const themeId = '1';
  const gridId = '2';

  const onToggleCollapse = () => {
    dispatch(toggleCollapseLong(!isCollapseLong));
  };

  const onChangeDays = () => {
    if (daysFieldRef && daysFieldRef.current) {
      const days = parseInt(daysFieldRef.current.value);
      dispatch(updateKeepDays(days));
    }
  };

  const onToggleTheme = () => {
    dispatch(toggleTheme(!isDarkTheme));
  };

  const onToggleGrid = () => {
    dispatch(toggleGrid(!isGrid));
  };

  const onClick = (e: MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
  };

  if (!isVisible) {
    return null;
  }

  const itemClass = block.getElement('item').toString();
  const labelClass = block.getElement('label').toString();
  const checkboxClass = block.getElement('checkbox').toString();

  return (
    <div className={block.toString()} onClick={onClick}>
      <div className={itemClass}>
        <input ref={themeFieldRef} id={themeId} className={checkboxClass} type="checkbox" checked={isDarkTheme} onChange={onToggleTheme} />
        <label className={labelClass} htmlFor={themeId}>Темная тема</label>
      </div>

      <div className={itemClass}>
        <input ref={collapseFieldRef} id={collapseId} className={checkboxClass} type="checkbox" checked={isCollapseLong} onChange={onToggleCollapse} />
        <label className={labelClass} htmlFor={collapseId}>Сворачивать длинные посты</label>
      </div>

      <div className={itemClass}>
        <input ref={gridFieldRef} id={gridId} className={checkboxClass} type="checkbox" checked={!isGrid} onChange={onToggleGrid} />
        <label className={labelClass} htmlFor={gridId}>Показывать лентой</label>
      </div>

      <div className={itemClass}>
        <div className={labelClass}>
          Удалять посты старше <input className={block.getElement('days').toString()} ref={daysFieldRef} type="text" value={keepDays} onChange={onChangeDays}/> дней
        </div>
      </div>
    </div>
  );
}

export default Settings;
