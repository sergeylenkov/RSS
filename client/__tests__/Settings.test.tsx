import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../src/store';
import Settings from '../src/components/settings/Settings';

test('Settings isVisible = true', () => {
  const render = renderer.create(
    <Provider store={store}>
      <Settings isVisible />
    </Provider>
  );
  expect(render).toMatchSnapshot();
});

test('Settings isVisible = false', () => {
  const render = renderer.create(
    <Provider store={store}>
      <Settings isVisible={false} />
    </Provider>
  );
  expect(render).toMatchSnapshot();
});
