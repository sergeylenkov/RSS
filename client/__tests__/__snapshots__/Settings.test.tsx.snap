// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings isVisible = false 1`] = `null`;

exports[`Settings isVisible = true 1`] = `
<div
  className="settings"
  onClick={[Function]}
>
  <div
    className="settings__item"
  >
    <input
      checked={false}
      className="settings__checkbox"
      id="1"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="settings__label"
      htmlFor="1"
    >
      Темная тема
    </label>
  </div>
  <div
    className="settings__item"
  >
    <input
      checked={true}
      className="settings__checkbox"
      id="0"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="settings__label"
      htmlFor="0"
    >
      Сворачивать длинные посты
    </label>
  </div>
  <div
    className="settings__item"
  >
    <input
      checked={true}
      className="settings__checkbox"
      id="2"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="settings__label"
      htmlFor="2"
    >
      Показывать лентой
    </label>
  </div>
  <div
    className="settings__item"
  >
    <div
      className="settings__label"
    >
      Удалять посты старше 
      <input
        className="settings__days"
        onChange={[Function]}
        type="text"
        value={30}
      />
       дней
    </div>
  </div>
</div>
`;
